File: backend/app/repositories/videos.py
from __future__ import annotations

from typing import Any


class VideosRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    def fetch_video_list_rows(self, *, start_iso: str | None, end_iso: str | None, limit: int) -> list[dict[str, Any]]:
        q = (
            self._supa.table("videos")
            .select(
                "video_id,title,channel,published_at,video_url,thumbnail_url,view_count,like_count,comment_count,duration_seconds,video_summaries(overall_explanation,sentiment)"
            )
            .order("published_at", desc=True)
            .limit(int(limit))
        )

        if start_iso is not None and end_iso is not None:
            q = q.gte("published_at", start_iso).lte("published_at", end_iso)

        resp = q.execute()
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_videos_basic_window(self, *, start_iso: str, end_iso: str, limit: int) -> list[dict[str, Any]]:
        resp = (
            self._supa.table("videos")
            .select("video_id,title,channel,published_at,video_url,thumbnail_url")
            .gte("published_at", start_iso)
            .lte("published_at", end_iso)
            .order("published_at", desc=True)
            .limit(int(limit))
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_summaries_for_video_ids(self, *, video_ids: list[str], limit: int = 5000) -> list[dict[str, Any]]:
        if not video_ids:
            return []
        resp = (
            self._supa.table("summaries")
            .select("video_id,ticker,summary")
            .in_("video_id", [str(x) for x in video_ids])
            .limit(int(limit))
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_video_row(self, *, video_id: str) -> dict[str, Any] | None:
        resp = (
            self._supa.table("videos")
            .select(
                "video_id,title,channel,published_at,video_url,thumbnail_url,view_count,like_count,comment_count,duration_seconds"
            )
            .eq("video_id", video_id)
            .limit(1)
            .execute()
        )
        row = resp.data[0] if resp.data else None
        return row if isinstance(row, dict) else None

    def fetch_video_summary_row(self, *, video_id: str) -> dict[str, Any] | None:
        resp = (
            self._supa.table("video_summaries")
            .select(
                "video_titles,published_at,summary_markdown,overall_explanation,movers,risks,opportunities,key_points,sentiment,events,model,summarized_at"
            )
            .eq("video_id", video_id)
            .limit(1)
            .execute()
        )
        row = resp.data[0] if resp.data else None
        return row if isinstance(row, dict) else None

    def fetch_video_ticker_rows(self, *, video_id: str, limit: int = 500) -> list[dict[str, Any]]:
        resp = self._supa.table("summaries").select("ticker").eq("video_id", video_id).limit(int(limit)).execute()
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

    def fetch_latest_per_ticker_summary_rows(self, *, video_id: str, limit: int = 500) -> list[dict[str, Any]]:
        resp = (
            self._supa.table("summaries")
            .select("ticker,summary,created_at")
            .eq("video_id", video_id)
            .order("created_at", desc=True)
            .limit(int(limit))
            .execute()
        )
        rows = resp.data or []
        return [r for r in rows if isinstance(r, dict)]

File: backend/internal/repo/videos.go
package repo

import (
	"fmt"

	"yunews/backend/internal/supabase"
)

// VideosRepository accesses the videos, video_summaries, and summaries tables.
type VideosRepository struct {
	db *supabase.Client
}

// NewVideosRepository creates a new VideosRepository.
func NewVideosRepository(db *supabase.Client) *VideosRepository {
	return &VideosRepository{db: db}
}

// FetchVideoListRows fetches videos with embedded video_summaries.
// If startISO/endISO are non-empty, filters by published_at range.
func (r *VideosRepository) FetchVideoListRows(startISO, endISO string, limit int) ([]map[string]interface{}, error) {
	q := r.db.From("videos").
		Select("video_id,title,channel,published_at,video_url,thumbnail_url,view_count,like_count,comment_count,duration_seconds,video_summaries(overall_explanation,sentiment)").
		Order("published_at", true).
		Limit(limit)

	if startISO != "" && endISO != "" {
		q = q.Gte("published_at", startISO).Lte("published_at", endISO)
	}

	var rows []map[string]interface{}
	if err := q.Execute(&rows); err != nil {
		return nil, fmt.Errorf("fetch video list rows: %w", err)
	}
	return rows, nil
}

// FetchVideosBasicWindow fetches video metadata within a published_at window.
func (r *VideosRepository) FetchVideosBasicWindow(startISO, endISO string, limit int) ([]map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("videos").
		Select("video_id,title,channel,published_at,video_url,thumbnail_url").
		Gte("published_at", startISO).
		Lte("published_at", endISO).
		Order("published_at", true).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch videos basic window: %w", err)
	}
	return rows, nil
}

// FetchSummariesForVideoIDs fetches ticker-level summaries for the given video IDs.
func (r *VideosRepository) FetchSummariesForVideoIDs(videoIDs []string, limit int) ([]map[string]interface{}, error) {
	if len(videoIDs) == 0 {
		return nil, nil
	}
	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("video_id,ticker,summary").
		In("video_id", videoIDs).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch summaries for video ids: %w", err)
	}
	return rows, nil
}

// FetchVideoRow fetches a single video row by video_id.
func (r *VideosRepository) FetchVideoRow(videoID string) (map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("videos").
		Select("video_id,title,channel,published_at,video_url,thumbnail_url,view_count,like_count,comment_count,duration_seconds").
		Eq("video_id", videoID).
		Limit(1).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch video row: %w", err)
	}
	if len(rows) == 0 {
		return nil, nil
	}
	return rows[0], nil
}

// FetchVideoSummaryRow fetches the video-level summary for a video_id.
func (r *VideosRepository) FetchVideoSummaryRow(videoID string) (map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("video_summaries").
		Select("video_titles,published_at,summary_markdown,overall_explanation,movers,risks,opportunities,key_points,sentiment,events,model,summarized_at").
		Eq("video_id", videoID).
		Limit(1).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch video summary row: %w", err)
	}
	if len(rows) == 0 {
		return nil, nil
	}
	return rows[0], nil
}

// FetchVideoTickerRows fetches the set of tickers associated with a video.
func (r *VideosRepository) FetchVideoTickerRows(videoID string, limit int) ([]map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("ticker").
		Eq("video_id", videoID).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch video ticker rows: %w", err)
	}
	return rows, nil
}

// FetchLatestPerTickerSummaryRows fetches per-ticker summaries for a video, newest first.
func (r *VideosRepository) FetchLatestPerTickerSummaryRows(videoID string, limit int) ([]map[string]interface{}, error) {
	var rows []map[string]interface{}
	err := r.db.From("summaries").
		Select("ticker,summary,created_at").
		Eq("video_id", videoID).
		Order("created_at", true).
		Limit(limit).
		Execute(&rows)
	if err != nil {
		return nil, fmt.Errorf("fetch latest per-ticker summary rows: %w", err)
	}
	return rows, nil
}
